File: communicationAPI/src/main/java/eu/cybergeiger/communication/StorableString.java
package eu.cybergeiger.communication;

import ch.fhnw.geiger.serialization.Serializer;
import ch.fhnw.geiger.serialization.SerializerHelper;
import ch.fhnw.geiger.totalcross.ByteArrayInputStream;
import ch.fhnw.geiger.totalcross.ByteArrayOutputStream;
import java.io.IOException;

/**
 * <p>A serializable yet simple String object.</p>
 */
public class StorableString implements Serializer {

  private static final long serialVersionUID = 142314912322198374L;

  private final String value;

  public StorableString() {
    value = "";
  }

  public StorableString(String value) {
    this.value = value;
  }

  public String toString() {
    return value;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    StorableString that = (StorableString) o;
    return (value == null && that.value == null) || (value != null && value.equals(that.value));
  }

  @Override
  public int hashCode() {
    return value.hashCode();
  }

  @Override
  public void toByteArrayStream(ByteArrayOutputStream out) throws IOException {
    SerializerHelper.writeString(out, this.value);
  }

  /**
   * <p>Reads objects from ByteArrayInputStream and stores them in map.</p>
   *
   * @param in ByteArrayInputStream to be used
   * @return the deserialized Storable String
   * @throws IOException if value cannot be read
   */
  public StorableString fromByteArrayStream(ByteArrayInputStream in) throws IOException {
    return new StorableString(SerializerHelper.readString(in));
  }
}

File: lib/src/utils/storable_string.dart
library geiger_api;

import 'package:geiger_localstorage/geiger_localstorage.dart';
import 'package:meta/meta.dart';

/// A serializable yet simple String object.
@immutable
class StorableString implements Serializer {
  static const int serialVersionUID = 142224912322198374;
  final String? value;

  const StorableString([this.value]);

  @override
  String toString() {
    return value ?? '';
  }

  @override
  bool operator ==(Object other) {
    return identical(this, other) ||
        (other is StorableString && value == other.value);
  }

  @override
  int get hashCode => Object.hash(StorableString, value.hashCode);

  @override
  void toByteArrayStream(ByteSink out) {
    SerializerHelper.writeLong(out, serialVersionUID);
    SerializerHelper.writeString(out, value);
  }

  static Future<StorableString> fromByteArrayStream(ByteStream in_) async {
    SerializerHelper.castTest('StorableString', serialVersionUID,
        await SerializerHelper.readLong(in_), 1);
    return StorableString(await SerializerHelper.readString(in_));
  }
}
